pkg/caveats/parameters.go

```diff
@@ -0,0 +1,62 @@
+package caveats
+
+import (
+	"fmt"
+
+	"github.com/rs/zerolog"
+
+	"github.com/authzed/spicedb/pkg/caveats/types"
+	core "github.com/authzed/spicedb/pkg/proto/core/v1"
+)
+
+// ConvertContextToParameters converts the given context into parameters of the types specified.
+// Returns a type error if type conversion failed.
+func ConvertContextToParameters(contextMap map[string]any, parameterTypes map[string]*core.CaveatTypeReference) (map[string]any, error) {
+	if len(contextMap) == 0 {
+		return nil, nil
+	}
+
+	if len(parameterTypes) == 0 {
+		return nil, fmt.Errorf("missing parameters for caveat")
+	}
+
+	converted := make(map[string]any, len(contextMap))
+
+	for key, value := range contextMap {
+		paramType, ok := parameterTypes[key]
```

Comment:
so if a context key-value is not known to the type system, we just ignore it. Technically we know it cannot be used in a caveat expression unless it's defined as a typed argument. I wonder if this is worth logging as "hey, you are sending stuff that won't be used"?

...or another way to look at this is: the context includes a superset of all the key-values required, because the evaluation may involve multiple caveats with disjoint set of arguments. which begs the question: what do we do about args that are named the same but have different types. I think we discussed forcing them to be globally unique? (i.e. no 2 caveat definitions can share an argument with the same name?)

Reply:
> so if a context key-value is not known to the type system, we just ignore it. Technically we know it cannot be used in a caveat expression unless it's defined as a typed argument. I wonder if this is worth logging as "hey, you are sending stuff that won't be used"?

We can't do that, because you might have some caveats using only a portion but you may not know which portion until the check occurs. Thus, we have to ignore them.

> 
> ...or another way to look at this is: the context includes a superset of all the key-values required, because the evaluation may involve multiple caveats with disjoint set of arguments. which begs the question: what do we do about args that are named the same but have different types. I think we discussed forcing them to be globally unique? (i.e. no 2 caveat definitions can share an argument with the same name?)

Or we make them all be scoped by caveat name. Either or, but we do need to decide

Reply:
Yeah - let's open an issue to keep track of this, we probably have to make that decision sooner than later